### PR TITLE
GH2275: Correctly reference runtime libraries from addins

### LINKED ISSUE #2275: System.Data.SqlClient in Cake.CoreCLR Assembly Load Error
### What You Are Seeing?
When referencing System.Data.SqlClient in my Cake file, I get the following exception when run from the global tool in a Docker container:

```
Assembly with same name is already loaded
```
When running from Windows (`build.ps1`), the reference isn't needed, and even when present, causes no issues.

I asked for assistance in the chat, and was directed here.

### What is Expected?
The Cake script should compile.

### What version of Cake are you using?
dotnet core global tool 0.30.0

### Are you running on a 32 or 64 bit system?
64-bit

### What environment are you running on?  Windows? Linux? Mac?
Docker (Linux on Windows) `microsoft/dotnet:2.1.400-sdk-stretch`

### Are you running on a CI Server?  If so, which one?
No

### How Did You Get This To Happen? (Steps to Reproduce)
In my `build.cake` file:
```
#addin "nuget:?package=System.Data.SqlClient&version=4.5.0"
#addin "nuget:?package=Cake.Docker&version=0.9.6"
#addin "nuget:?package=dbup-core&version=4.1.0"
#addin "nuget:?package=dbup-sqlserver&version=4.1.0"
```

In my Docker file:

```
FROM microsoft/dotnet:2.1.400-sdk-stretch AS build-env

WORKDIR /app
COPY ./ ./
ENV PATH="/root/.dotnet/tools:${PATH}"
RUN dotnet tool install -g Cake.Tool --version 0.30.0
RUN dotnet cake build.cake
```

### Output Log

~~~sh
root@dc8cbcdce3b2:/app# dotnet cake build.cake -v=Diagnostic
Module directory does not exist.
NuGet.config not found.
Analyzing build script...
Analyzing /app/build.cake...
Processing build script...
Installing tools...
Retrieving package 'ReportGenerator 4.0.0-rc4' from '/root/.nuget/packages/'.
Package ReportGenerator.4.0.0-rc4 has already been installed.
Successfully installed 'ReportGenerator 4.0.0-rc4' to /app/tools
Executing nuget actions took 68.73 ms
Installing addins...
Found package 'System.Data.SqlClient 4.5.0' in '/app/tools/Addins'.
Package System.Data.SqlClient.4.5.0 has already been installed.
Successfully installed 'System.Data.SqlClient 4.5.0' to /app/tools/Addins
Executing nuget actions took 2.18 ms
The addin System.Data.SqlClient will reference System.Data.SqlClient.dll.
The addin System.Data.SqlClient will reference System.Data.SqlClient.dll.
The addin System.Data.SqlClient will reference System.Data.SqlClient.dll.
Found package 'Cake.Docker 0.9.6' in '/app/tools/Addins'.
Package Cake.Docker.0.9.6 has already been installed.
Successfully installed 'Cake.Docker 0.9.6' to /app/tools/Addins
Executing nuget actions took 0.46 ms
The addin Cake.Docker will reference Cake.Docker.dll.
Found package 'dbup-core 4.1.0' in '/app/tools/Addins'.
Package dbup-core.4.1.0 has already been installed.
Successfully installed 'dbup-core 4.1.0' to /app/tools/Addins
Executing nuget actions took 0.65 ms
The addin dbup-core will reference dbup-core.dll.
Found package 'dbup-sqlserver 4.1.0' in '/app/tools/Addins'.
Package dbup-sqlserver.4.1.0 has already been installed.
Successfully installed 'dbup-sqlserver 4.1.0' to /app/tools/Addins
Executing nuget actions took 0.57 ms
The addin dbup-sqlserver will reference dbup-sqlserver.dll.
Verifying assembly 'System.Data.SqlClient, Version=4.4.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
Error: System.IO.FileLoadException: Assembly with same name is already loaded
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromPath(IntPtr ptrNativeAssemblyLoadContext, String ilPath, String niPath, ObjectHandleOnStack retAssembly)
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at Cake.Core.Reflection.AssemblyLoader.Load(FilePath path, Boolean verify) in C:\projects\cake\src\Cake.Core\Reflection\AssemblyLoader.cs:line 31
   at Cake.Core.Scripting.ScriptRunner.Run(IScriptHost host, FilePath scriptPath, IDictionary`2 arguments) in C:\projects\cake\src\Cake.Core\Scripting\ScriptRunner.cs:line 171
   at Cake.Commands.BuildCommand.Execute(CakeOptions options) in C:\projects\cake\src\Cake\Commands\BuildCommand.cs:line 41
   at Cake.CakeApplication.Run(CakeOptions options) in C:\projects\cake\src\Cake\CakeApplication.cs:line 45
   at Cake.Program.Main() in C:\projects\cake\src\Cake\Program.cs:line 73
~~~

### PR DESCRIPTION
**Pull request by mholo65**
- Fixes #2275

~~Don't know if this is the right direction to go. Basically we *should* load the runtime libs based on the current RID if running .NET Core. Problem is there's a whole lot of them (see [list or runtimes](https://github.com/dotnet/corefx/blob/7e56d8bd8b224d834b782a98bfd7035f71c3f547/pkg/Microsoft.NETCore.Platforms/runtime.json)  and [docs](https://docs.microsoft.com/en-us/dotnet/core/rid-catalog)).~~

~~@daveaglick do you have anything to add? Maybe we should consult our friends at NuGet?~~

Went with the route of correctly findint runtime assemblies and loading them, including native libraries.

### CODE REVIEW

File: src/Cake.NuGet/NuGetContentResolver.cs [hunk 1]
```diff
@@ -63,15 +63,17 @@ private IReadOnlyCollection<IFile> GetAddinAssemblies(DirectoryPath path, Packag
             var provider = DefaultFrameworkNameProvider.Instance;
             var current = NuGetFramework.Parse(_environment.Runtime.BuiltFramework.FullName, provider);
 
-            // Get all ref assemblies.
+            // Get all ref and runtime assemblies.
             var refAssemblies = _globber.GetFiles(path.FullPath + "/ref/**/*.dll");
+            var runtimeAssemblies = _globber.GetFiles(path.FullPath + "/runtimes/**/*.dll");
```
Comment (devlead) [inline src/Cake.NuGet/NuGetContentResolver.cs]: As `IsClrAssembly` looks at the files maybe we should reduce files it should even look at first, something like 
```csharp
GetFiles(path, package, new[] { path.FullPath + "/**/*.dll" })
    .Except(refAssemblies + runtimeAssemblies, pathComparer)
```

And maybe even change `refAssemblies + runtimeAssemblies` to something like
```csharp
var excludedAssemblies = _globber.GetFiles(path.FullPath + "/{ref,runtimes}/**/*.dll");
```

Not tested if it compiles, but gist of it that it could reduce number of directory scans.

File: src/Cake.NuGet/NuGetContentResolver.cs [hunk 2]
```diff
@@ -63,15 +63,17 @@ private IReadOnlyCollection<IFile> GetAddinAssemblies(DirectoryPath path, Packag
...
 
             // Get all candidate files.
             var pathComparer = PathComparer.Default;
             var assemblies = GetFiles(path, package, new[] { path.FullPath + "/**/*.dll" })
                 .Where(file => !"Cake.Core.dll".Equals(file.Path.GetFilename().FullPath, StringComparison.OrdinalIgnoreCase)
                                && file.IsClrAssembly()
-                               && !refAssemblies.Contains(file.Path, pathComparer))
+                               && !refAssemblies.Contains(file.Path, pathComparer)
```
Comment (devlead) [inline src/Cake.NuGet/NuGetContentResolver.cs]: See above.

File: src/Cake.NuGet/NuGetContentResolver.cs [hunk 3]
```diff
@@ -63,15 +63,17 @@ private IReadOnlyCollection<IFile> GetAddinAssemblies(DirectoryPath path, Packag
...
+                               && !runtimeAssemblies.Contains(file.Path, pathComparer))
```
Comment (devlead) [inline src/Cake.NuGet/NuGetContentResolver.cs]: See above.

**Comment by mholo65** [comment]
As per https://github.com/cake-build/cake/issues/2275#issuecomment-510231758 I think this approach is wrong. We actually WANT to load the `runtime` libs is present and not the one in lib. Loading the runtime ones is tricky tough as we need to find the best RID... Don't know if there's a lib for that. Anyhow, do not merge this.

File: src/Cake.NuGet/NuGetContentResolver.cs (hunk 1 above)
Comment (patriksvensson) [reply src/Cake.NuGet/NuGetContentResolver.cs]: Sounds like a good idea!

**Comment by mholo65** [comment]
OMG! Got this working on CoreCLR. Here's a script.

```
#addin nuget:?package=System.Data.SqlClient&version=4.6.1&loaddependencies=true

using System.Data.SqlClient;

Task("Default")
.Does(() =>
{
    using(var connection = new SqlConnection(@"Data Source=.\SQLEXPRESS;Initial Catalog=master;Integrated Security=True"))
    {
        connection.Open();

        using (var cmd = new SqlCommand("SELECT name from sys.databases", connection))
        {
            using (var dr = cmd.ExecuteReader())
            {
                while (dr.Read())
                {
                    System.Console.WriteLine(dr[0].ToString());
                }
            }
        }
    }
});

RunTarget("Default");
```

Now I'll just tidy up the code.

File: src/Cake.Core/Polyfill/EnvironmentHelper.cs [hunk 4]
```diff
@@ -119,7 +120,10 @@ public static Runtime GetRuntime()
         public static FrameworkName GetBuiltFramework()
         {
 #if NETCORE
-            return new FrameworkName(".NETStandard,Version=v2.0");
+            return new FrameworkName(Assembly
+                .GetEntryAssembly()?
+                .GetCustomAttribute<TargetFrameworkAttribute>()?
+                .FrameworkName ?? ".NETStandard,Version=v2.0");
```
Comment (mholo65) [inline src/Cake.Core/Polyfill/EnvironmentHelper.cs]: This allows us to get `.NETCoreApp,Version=v2.1` for `Cake.Tool` and `.NETCoreApp,Version=v2.0` for `Cake.CoreCLR`.

File: src/Cake.NuGet/Cake.NuGet.csproj [hunk 5]
```diff
@@ -26,5 +26,16 @@
...
     <PackageReference Include="NuGet.Common" Version="5.0.0" />
     <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
+    <PackageReference Include="Microsoft.DotNet.PlatformAbstractions" Version="2.0.0" />
   </ItemGroup>
+
+  <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' ">
+    <PackageReference Include="Microsoft.NETCore.Platforms" Version="2.2.2" GeneratePathProperty="true">
+      <PrivateAssets>All</PrivateAssets>
+    </PackageReference>
+    <EmbeddedResource  Include="$(PkgMicrosoft_NETCore_Platforms)\runtime.json">
```
Comment (mholo65) [inline src/Cake.NuGet/Cake.NuGet.csproj]: I'm embedding [this](https://github.com/dotnet/corefx/blob/master/pkg/Microsoft.NETCore.Platforms/runtime.json) in order to build the runtime graph.

File: src/Cake.NuGet/NuGetContentResolver.cs [hunk 6]
```diff
@@ -56,82 +60,48 @@ private IReadOnlyCollection<IFile> GetAddinAssemblies(DirectoryPath path, Packag
...
             {
-                // Get relative path.
-                var relative = path.GetRelativePath(assembly.Path);
-                var framework = ParseFromDirectoryPath(current, relative.GetDirectory());
-                if (!mapping.ContainsKey(framework))
-                {
-                    mapping.Add(framework, new List<FilePath>());
-                }
-                mapping[framework].Add(assembly.Path);
+                runtimeGraph = JsonRuntimeFormat.ReadRuntimeGraph(stream);
```
Comment (mholo65) [inline src/Cake.NuGet/NuGetContentResolver.cs]: Make this static lazy

**Comment by mholo65** [comment]
@cake-build/cake-team ready for review!

File: src/Cake.Core/Polyfill/EnvironmentHelper.cs [hunk 7]
```diff
@@ -123,6 +123,7 @@ public static Runtime GetRuntime()
         public static FrameworkName GetBuiltFramework()
         {
 #if NETCORE
+
```
Comment (devlead) [inline src/Cake.Core/Polyfill/EnvironmentHelper.cs]: Guess this isn't needed?

File: src/Cake.NuGet/NuGetContentResolver.cs (hunk 6 above)
Comment (devlead) [reply src/Cake.NuGet/NuGetContentResolver.cs]: Is this something you want to address?

**Comment by mholo65** [comment]
@patriksvensson @devlead I added some tests and fixed up code. Please re-review.

File: src/Cake.NuGet.Tests/Fixtures/NuGetContentResolverFixture.cs [hunk 8]
```diff
@@ -28,6 +28,7 @@ protected NuGetContentResolverFixture(string framework, Runtime runtime)
             Environment = FakeEnvironment.CreateUnixEnvironment();
             Environment.Runtime.BuiltFramework = new FrameworkName(framework);
             Environment.Runtime.Runtime = runtime;
+            Environment.Runtime.IsCoreClr = runtime == Runtime.CoreClr;
```
Comment (mholo65) [inline src/Cake.NuGet.Tests/Fixtures/NuGetContentResolverFixture.cs]: Why is it that the `FakeEnvironment` has a setter for this? Wouldn't it be beneficial to just check `Runtime == Runtime.CoreClr`?

I can fix it in a separate PR.

File: src/Cake.Core/Polyfill/EnvironmentHelper.cs [hunk 9]
```diff
@@ -119,7 +119,28 @@ public static Runtime GetRuntime()
...
+            var runtimeBasePathLength = assemblyPath.IndexOf(microsoftNetCoreApp) + microsoftNetCoreApp.Length + 1;
+            var netCoreAppVersion = string.Concat(assemblyPath.Skip(runtimeBasePathLength).Take(3));
+            if (string.IsNullOrEmpty(netCoreAppVersion))
+            {
+                return NetStandardFramework;
+            }
+
+            return netCoreAppFramwork = Version.TryParse(netCoreAppVersion, out var version)
+                                            ? new FrameworkName(".NETCoreApp", version)
+                                            : NetStandardFramework;
```
Comment (mholo65) [inline src/Cake.Core/Polyfill/EnvironmentHelper.cs]: Rebase gone bad, or why is this change here? Didn't @devlead write this in another commit?

**Comment by mholo65** [comment]
> 
> 
> There seems to be failing tests on at least GitHub actions and macOs (Don't know about Windows/Linux as those were aborted because the failing macOS.
> 
> ```diff
> - Cake.NuGet.Tests.Unit.NuGetContentResolverTests+Addins.Should_Return_Native_Runtimes_Assemblies_If_CoreCLR [FAIL]
> ```

Weird as those don't fail on AzDo Mac? @devlead you got a Mac, can you reproduce locally?

**Comment by devlead** [comment]
> Weird as those don't fail on AzDo Mac? @devlead you got a Mac, can you reproduce locally?

I've rebased so GH Actions builds on 3.1, and it still fails at same point.

Haven't tried on my Machine yet. but interestingly macOS on Bitrise and Azure DevOps test fine.

Wonder if it's a roll forward issue, as GH Actions has fewer versions of .NET Core.

There's something interesting introduced with this PR as other PRs doesn't fail.

**Comment by devlead** [comment]
Ok these tests fail on my Mac too, so able to repro
```diff
- Cake.NuGet.Tests.Unit.NuGetContentResolverTests+Addins.Should_Return_Native_Runtimes_Assemblies_If_CoreCLR [FAIL]
- Cake.NuGet.Tests.Unit.NuGetContentResolverTests+Addins.Should_Return_Runtimes_Assemblies_If_CoreCLR [FAIL]
```

**Comment by devlead** [comment]
Some notes, on my Mac I get

### Should_Return_Runtimes_Assemblies_If_CoreCLR
* tfm: ".NETCoreApp,Version=v2.0"
* rid: osx.10.15-x64
* assemblies: 3
* managedAssemblies: null
* nativeAssemblies: null
* result: 0

### Should_Return_Runtimes_Assemblies_If_CoreCLR
* tfm: ".NETCoreApp,Version=v2.0"
* rid: osx.10.15-x64
* assemblies: 2
* managedAssemblies: null
* nativeAssemblies: null
* result: 0

Updated to NuGet client libs 5.4.0, with same results.

Maybe something with newer macOS version?

**Comment by devlead** [comment]
Ok time for bed but think I found issue, after updating client libs and Microsoft.NETCore.Platforms it started working, as it got a newer runtime.json containing latest macOS rid.

Pushed new commit and looks like tests now pass on GH action.

@mholo65 please take a look.

**Comment by devlead** [comment]
@mholo65 your changes have been merged, thanks for your contribution 👍
